router/usecase/pools/routable_cw_orderbook_pool.go

```diff
@@ -0,0 +1,188 @@
+
+	"github.com/osmosis-labs/sqs/domain"
+	"github.com/osmosis-labs/sqs/sqsdomain"
+
+	"github.com/osmosis-labs/osmosis/osmomath"
+	clmath "github.com/osmosis-labs/osmosis/v25/x/concentrated-liquidity/math"
+	cwpoolmodel "github.com/osmosis-labs/osmosis/v25/x/cosmwasmpool/model"
+	poolmanagertypes "github.com/osmosis-labs/osmosis/v25/x/poolmanager/types"
+)
+
+var _ sqsdomain.RoutablePool = &routableOrderbookPoolImpl{}
+
+type routableOrderbookPoolImpl struct {
+	ChainPool     *cwpoolmodel.CosmWasmPool     "json:\"pool\""
+	Balances      sdk.Coins                     "json:\"balances\""
+	TokenOutDenom string                        "json:\"token_out_denom\""
+	TakerFee      osmomath.Dec                  "json:\"taker_fee\""
+	SpreadFactor  osmomath.Dec                  "json:\"spread_factor\""
+	TickModel     *sqsdomain.OrderbookTickModel "json:\"orderbook_tick_model\""
+	QuoteDenom    string                        "json:\"quote_denom\""
+	BaseDenom     string                        "json:\"base_denom\""
+}
+
+// GetId implements sqsdomain.RoutablePool.
+func (r *routableOrderbookPoolImpl) GetId() uint64 {
+	return r.ChainPool.PoolId
+}
+
+// GetPoolDenoms implements sqsdomain.RoutablePool.
+func (r *routableOrderbookPoolImpl) GetPoolDenoms() []string {
+	return r.Balances.Denoms()
+}
+
+// GetType implements sqsdomain.RoutablePool.
+func (*routableOrderbookPoolImpl) GetType() poolmanagertypes.PoolType {
+	return poolmanagertypes.CosmWasm
+}
+
+// GetSpreadFactor implements sqsdomain.RoutablePool.
+func (r *routableOrderbookPoolImpl) GetSpreadFactor() math.LegacyDec {
+	return r.SpreadFactor
+}
+
+func (r *routableOrderbookPoolImpl) CalculateTokenOutByTokenIn(ctx context.Context, tokenIn sdk.Coin) (sdk.Coin, error) {
+	poolType := r.GetType()
+
+	// Esnure that the pool is a cosmwasm pool
+	if poolType != poolmanagertypes.CosmWasm {
+		return sdk.Coin{}, domain.InvalidPoolTypeError{PoolType: int32(poolType)}
+	}
+
+	// Get the expected order direction
+	direction, err := r.GetDirection(tokenIn.Denom, r.TokenOutDenom)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	tickIdx, err := r.GetStartTickIndex(direction)
+	if err != nil || tickIdx == -1 {
+		return sdk.Coin{}, err
+	}
+
+	amountOutTotal := osmomath.ZeroBigDec()
+	amountInRemaining := osmomath.BigDecFromSDKInt(tokenIn.Amount)
+
+	// ASSUMPTION: Ticks are ordered
+	for amountInRemaining.GT(zeroBigDec) {
+		if tickIdx >= len(r.TickModel.TickStates) || tickIdx < 0 {
+			return sdk.Coin{}, sqsdomain.OrderbookNotEnoughLiquidityToCompleteSwapError{PoolId: r.GetId(), AmountIn: tokenIn}
+		}
+		tick := r.TickModel.TickStates[tickIdx]
+
+		if direction == sqsdomain.ASK {
+			tickIdx++
+		} else if direction == sqsdomain.BID {
+			tickIdx--
+		} else {
+			return sdk.Coin{}, sqsdomain.OrderbookPoolInvalidDirectionError{Direction: direction}
+		}
+
+		// Calculate the price for the current tick
+		tickPrice, err := clmath.TickToPrice(tick.TickId)
+		if err != nil {
+			return sdk.Coin{}, err
+		}
```

Comment:
Is our order book tick-to-price conversion the same as CL?

Reply:
Yes, but reimplemented in rust so there might be minor differences due to precision diffs